File: solutions/ruby/simple-cipher/simple_cipher.rb
class Cipher
  private

  attr_accessor :shifts
  attr_writer :key

  LETTERS = ('a'..'z').to_a.freeze
  ORD_A = 'a'.ord

  def initialize(key = nil)
    key = (0..100).map { (rand(LETTERS.length) + ORD_A).chr }.join if key.nil?

    raise ArgumentError, 'Missing key' if key == ''
    raise ArgumentError, 'Invalid key' unless key.chars.to_a.all? { |letter| LETTERS.include?(letter) }

    self.key = key
    self.shifts = key.chars.map { |letter| LETTERS.index(letter) }
  end

  def encode_with(text, shifts)
    text.chars.zip(shifts.cycle).map { |letter, shift| ((letter.ord - ORD_A + shift) % 26 + ORD_A).chr }.join
  end

  public

  attr_reader :key

  def encode(text)= encode_with(text, shifts)
  def decode(text)= encode_with(text, shifts.map(&:-@))
end

File: solutions/python/simple-cipher/simple_cipher.py
"""Simple cipher."""

import itertools
import random
import string

ORD_A = ord("a")


def add(char_a: str, char_b: str, reverse: bool) -> str:
    """Add two chars together. Or subtract, if reverse."""
    val = ord(char_a) - ORD_A
    dist = ord(char_b) - ORD_A
    if reverse:
        dist = -dist
    res = (val + dist) % 26
    return chr(res + ORD_A)


class Cipher:
    """Simple cipher."""

    def __init__(self, key: str | None = None):
        self.key = key or "".join(random.choices(string.ascii_lowercase, k=100))

    def encode(self, text: str, reverse: bool = False) -> str:
        """Return an encoded string using the key."""
        return "".join(add(char, key, reverse) for char, key in zip(text, itertools.cycle(self.key)))

    def decode(self, text: str) -> str:
        """Return a decoded string using the key."""
        return self.encode(text, reverse=True)
